File: Files/stockdash2-3/frontend/src/types/stock.ts
export interface Quote {
  ticker: string
  price: number | null
  change: number | null
  change_pct: number | null
  high: number | null
  low: number | null
  open: number | null
  prev_close: number | null
  timestamp: number | null
}

export interface CandlePoint {
  date: string
  timestamp: number
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface CandlesResponse {
  ticker: string
  range: string
  resolution: string
  candles: CandlePoint[]
}

export interface SymbolResult {
  ticker: string | null
  description: string | null
  type: string | null
  exchange: string | null
}

export interface SearchResponse {
  query: string
  results: SymbolResult[]
}

export interface NewsArticle {
  id: number | null
  headline: string | null
  summary: string | null
  source: string | null
  url: string | null
  image: string | null
  sentiment: 'positive' | 'neutral' | 'negative'
  published_at: string | null
}

export interface CompanyProfile {
  ticker: string
  company_name: string | null
  sector: string | null
  market_cap: number | null
  logo_url: string | null
  exchange: string | null
  ipo_date: string | null
  website: string | null
  country: string | null
  currency: string | null
}

export interface BasicFinancials {
  ticker: string
  week_52_high: number | null
  week_52_low: number | null
  beta: number | null
  pe_ratio: number | null
  eps: number | null
  dividend_yield: number | null
}

export interface NormalizedPoint {
  date: string
  timestamp: number
  pct_return: number
  close: number
}

export interface CompareSeries {
  ticker: string
  color: string
  start_price: number
  end_price: number
  pct_change: number
  points: NormalizedPoint[]
}

export interface CompareResponse {
  tickers: string[]
  range: string
  series: CompareSeries[]
}

export type TimeRange = '1D' | '1W' | '1M' | '3M' | '1Y' | '5Y'
export type ChartType = 'line' | 'candlestick'

File: Files/stockdash2-8/backend/app/schemas/stock.py
from datetime import datetime
from typing import Any
from pydantic import BaseModel, Field


# ─── Quote ───────────────────────────────────────────────────────────────────

class QuoteResponse(BaseModel):
    ticker: str
    price: float | None
    change: float | None
    change_pct: float | None
    high: float | None
    low: float | None
    open: float | None
    prev_close: float | None
    timestamp: int | None


# ─── Candles ─────────────────────────────────────────────────────────────────

class CandlePoint(BaseModel):
    date: str
    timestamp: int
    open: float
    high: float
    low: float
    close: float
    volume: float


class CandlesResponse(BaseModel):
    ticker: str
    range: str
    resolution: str
    candles: list[CandlePoint]


# ─── Compare (multi-stock) ───────────────────────────────────────────────────

class NormalizedSeriesPoint(BaseModel):
    date: str
    timestamp: int
    pct_return: float   # % return from start date, e.g. 12.4 = +12.4%
    close: float


class CompareSeriesItem(BaseModel):
    ticker: str
    color: str          # Assigned chart color hex
    start_price: float
    end_price: float
    pct_change: float
    points: list[NormalizedSeriesPoint]


class CompareResponse(BaseModel):
    tickers: list[str]
    range: str
    series: list[CompareSeriesItem]


# ─── Symbol Search ───────────────────────────────────────────────────────────

class SymbolResult(BaseModel):
    ticker: str | None
    description: str | None
    type: str | None
    exchange: str | None


class SearchResponse(BaseModel):
    query: str
    results: list[SymbolResult]


# ─── News ─────────────────────────────────────────────────────────────────────

class NewsArticle(BaseModel):
    id: int | None = None
    headline: str | None
    summary: str | None
    source: str | None
    url: str | None
    image: str | None
    sentiment: str = "neutral"
    published_at: str | None


class NewsResponse(BaseModel):
    ticker: str
    articles: list[NewsArticle]


# ─── Company Profile ─────────────────────────────────────────────────────────

class CompanyProfileResponse(BaseModel):
    ticker: str
    company_name: str | None
    sector: str | None
    market_cap: float | None
    logo_url: str | None
    exchange: str | None
    ipo_date: str | None
    website: str | None
    country: str | None
    currency: str | None


# ─── Financials ──────────────────────────────────────────────────────────────

class BasicFinancialsResponse(BaseModel):
    ticker: str
    week_52_high: float | None = Field(None, alias="52_week_high")
    week_52_low: float | None = Field(None, alias="52_week_low")
    beta: float | None
    pe_ratio: float | None
    eps: float | None
    revenue_per_share: float | None
    dividend_yield: float | None

    model_config = {"populate_by_name": True}


# ─── Sentiment ───────────────────────────────────────────────────────────────

class SentimentResponse(BaseModel):
    ticker: str
    score: float             # -1.0 (bearish) to 1.0 (bullish)
    label: str               # positive / neutral / negative
    post_volume: int
    window_hours: int
    computed_at: str | None


class SentimentHistoryPoint(BaseModel):
    score: float
    label: str
    post_volume: int
    computed_at: str


class SentimentHistoryResponse(BaseModel):
    ticker: str
    history: list[SentimentHistoryPoint]


# ─── Market Overview ──────────────────────────────────────────────────────────

class MarketOverviewItem(BaseModel):
    ticker: str
    company_name: str | None
    sector: str | None
    price: float | None
    change: float | None
    change_pct: float | None
    volume: float | None
    market_cap: float | None


class MarketOverviewResponse(BaseModel):
    items: list[MarketOverviewItem]
    updated_at: str


# ─── Health ───────────────────────────────────────────────────────────────────

class HealthResponse(BaseModel):
    status: str
    db: str
    redis: str
    version: str = "1.0.0"
